[fetch_binary_drivers] new [fetch_drivers] package

moriarty opened the pull request:
This is a new package providing the "fetch_drivers" which was previously
built internally and distributed as a binary only debian package.

We will officially announce when the upgrade from Indigo->Melodic is
ready. _**This is not yet the announcement**_.

18.04 and ROS Melodic on the Fetch Research Platforms is still being
tested.

The process isn't straight foward as there are calibration files
which you will want to back-up before the upgrade.

This `fetch_drivers` package also won't start like it use to because of
the change from `upstart` to `systemd`.
That requires `fetch_system_config` which will be open sourced soon.

This `fetch_binary_drivers` package is part of a plan to distribute
updates and upgrades to Fetch Research Platform customers faster in the future.

This relates to fetchrobotics/fetch_ros#63

1. An officiall announcement will be made when everything is ready.
2. We will update and document the process on:
    https://docs.fetchrobotics.com/
3. We will also announce via our mailing list, ros discourse and post on:
    https://opensource.fetchrobotics.com/

erelson committed:
System config debian components for ROS Melodic

Ported from internal Fetch Robotics code used for ROS Indigo/Ubuntu
14.04.  Updated to support ROS Melodic/Ubuntu 18.04.

Some key changes:
- Change from using upstart to using systemd (expect version >= 236)
- Correspondingly logs such as /var/log/upstart/robot.log are now
  available in /var/log/ros/robot.log (as well as via journalctl -u
  robot)
- Manually creates udev rules and updates grub boot arguments to ensure
  consistent network device naming across different hardware. Uses
  netplan to set static IP for internal robot communications.
- Sixad replaced with ps3joy (hopefully a better solution will be found)
- Slightly better logrotate rules
- Removed former joystick_monitor and soundplay upstarts

moriarty committed:
[fetch_binary_drivers] new [fetch_drivers] package

This is a new package providing the "fetch_drivers" which was previously
built internally and distributed as a binary only debian package.

We will officially announce when the upgrade from Indigo->Melodic is
ready.

18.04 and ROS Melodic on the Fetch Research Platforms is still being
tested.

The process isn't straight foward as there are calibration files
which you will want to back-up before the upgrade.

This `fetch_drivers` package also won't start like it use to because of
the change from `upstart` to `systemd`.
That requires `fetch_system_config` which will be open sourced soon.

This `fetch_binary_drivers` package is part of a plan to distribute
updates and upgrades to Fetch Research Platform customers faster in the future.

This relates to fetchrobotics/fetch_ros#63

1. An officiall announcement will be made when everything is ready.
2. We will update and document the process on:
    https://docs.fetchrobotics.com/
3. We will also announce via our mailing list, ros discourse and post on:
    https://opensource.fetchrobotics.com/

moriarty committed:
[fetch_system_config] fix README

TODO: should this be catkinized

erelson committed:
Fixup system config debian

erelson committed:
Fixup .gitignore

moriarty commented on fetch_binary_drivers/CMakeLists.txt:
```diff
@@ -0,0 +1,82 @@
+catkin_package(
+#  INCLUDE_DIRS include
+#  LIBRARIES fetch_binary_drivers
+#  CATKIN_DEPENDS other_catkin_pkg
```
TODO: the internal driver CATKIN_DEPENDS are slightly different between this version and the internal version (for example: no sick microscan3 safety scanners) so I need to sort through the list and include only what we need.

moriarty commented on fetch_binary_drivers/CMakeLists.txt:
```diff
@@ -0,0 +1,82 @@
+#  INCLUDE_DIRS include
+#  LIBRARIES fetch_binary_drivers
+#  CATKIN_DEPENDS other_catkin_pkg
+#  DEPENDS system_lib
```
I'm sure there are system dependencies. the binary driver is a shared object not a static one.

moriarty committed:
[fetch_drivers] add more dependencies

All these dependencies are exec_dependencies because this
package doesn't compile anything.

They're also in the CMakeLists.txt in the catkin_package and find_package to be safe.

moriarty reviewed (commented):
I will add additional dependencies just to be safe

cjds commented on fetch_binary_drivers/README.md:
```diff
@@ -0,0 +1,43 @@
+
+We will announce to our customers when we're officially ready and supporting melodic.
+
+See https://docs.fetchrobotics.com for more information.
```
Link here

cjds commented on fetch_binary_drivers/package.xml:
```diff
@@ -0,0 +1,66 @@
+    This package, is a cmake/make only package which installs the binaries for the drivers and firmware.
+  </description>
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
```
Are these comments intentionally left in

moriarty replied:
They're just left over from the catkin generate package command. I think it was just reminding me to add the maintainers.

cjds commented on fetch_system_config/debian/compat:
```diff
@@ -0,0 +1 @@
+9
```
You probably want 11?
```
Package: debhelper (9.20131227ubuntu1)
```
even for 14.04

erelson replied:
My 14.04 desktop (still on 3.x kernel... nice and safe...) has 9.x debhelper installed. I don't think there's a reason to bump this requirement presently. And it sounds like the plan is we'll move to using bloom for this later anyways.

cjds commented on fetch_system_config/debian/control:
```diff
@@ -0,0 +1,38 @@
+         openssh-server,
+         ros-melodic-ros
+Description: Configuration for Fetch on Melodic
+  This package will start ROS Melodic at bootup
```
What does this mean?

erelson replied:
Very brief vague statement, but nominally accurate. It installs roscore + robot systemd units which run at computer power-on. I'll reword this probably.

cjds commented on fetch_system_config/root/lib/udev/rules.d/99-logitechF710.rules:
```diff
@@ -0,0 +1,2 @@
+KERNEL=="js?", SUBSYSTEM=="input", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c219", SYMLINK+="logitechF710"
```
Why are we adding these rules unless we are giving them a logitech?

erelson replied:
This rule was already in there. We've enabled logitech controller usage historically and so it seems reasonable to continue doing so. (I would need to check if research stack has the teleop changes needed for this to work though)

cjds commented on fetch_system_config/root/etc/pm/power.d/wireless:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+# This is essential for reliable robot wifi
+/sbin/iwconfig wlan0 power off
```
Oh boy!!!

erelson replied:
I need to reinvestigate this.

erelson replied:
This is supposed to turn of power managment of the wifi card, and works in 14.04 (checked via output of `iwconfig` but isn't working on Fetch1 with 18.04.

Looks like the new way to make this work, (and verified) is to change wifi.powersave = 3 to = 2 in /etc/NetworkManager/conf.d/default-wifi-powersave-on.conf :+1:

cjds commented on fetch_binary_drivers/CMakeLists.txt:
```diff
@@ -0,0 +1,98 @@
+     actionlib_msgs
+     diagnostic_msgs
+     fetch_driver_msgs
+     fetch_auto_dock_msgs
```
I can't find a reference in the drivers to auto_dock

moriarty replied:
@cjds it might be stripped out of the research platform drivers.

```
amoriarty at SW44 in ~/ros/melodic/fetch-private/src/fetch_drivers on master
$ grep -rl auto_dock_msgs .
./CMakeLists.txt
./include/fetch_drivers/logpro/client.h
./package.xml
./src/logpro_client.cpp
```

cjds commented on fetch_binary_drivers/Makefile.tarball:
```diff
@@ -0,0 +1,19 @@
+all: binary_driver
+
+# TODO: There are some more things to automate.
```
More specific?

moriarty replied:
There are too many to list. I had several TODOs and I just combined them into one generic one.
I see your comments also say `Why not use DRIVER_VERSION` and that's basically when I rolled back and just used the hardcoded path for now.
But it goes into the comments on another comment were we'll need to setup more build/test automation.

For now we're just pulling `DRIVER_VERSION = 0.8.0` and until we have a more automated pipeline we'll leave it at this version and just be sure to rebuild that version of our drivers against the latest melodic especially at the time of a sync.

cjds commented on fetch_binary_drivers/Makefile.tarball:
```diff
@@ -0,0 +1,19 @@
+
+# TODO: There are some more things to automate.
+DRIVER_VERSION = 0.8.0
+TARBALL = build/fetch-drivers-0.8.0.tar.gz
```
Why not use DRIVER_VERSION

cjds commented on fetch_binary_drivers/Makefile.tarball:
```diff
@@ -0,0 +1,19 @@
+# TODO: There are some more things to automate.
+DRIVER_VERSION = 0.8.0
+TARBALL = build/fetch-drivers-0.8.0.tar.gz
+TARBALL_URL = http://packages.fetchrobotics.com/binaries/fetch-drivers-0.8.0.tar.gz
```
Ditto as above

cjds commented on fetch_binary_drivers/Makefile.tarball:
```diff
@@ -0,0 +1,19 @@
+all: binary_driver
```
Why non-standard name?

cjds commented on fetch_binary_drivers/README.md:
```diff
@@ -0,0 +1,43 @@
+We have two Fetch Research Platforms. Commonly known as Fetch and Freight. Fetch is the one with the arm.
+The drivers and firmware in this package are for both.
+
+https://docs.fetchrobotics.com/
```
Link

cjds commented on fetch_binary_drivers/README.md:
```diff
@@ -0,0 +1,43 @@
+
+This is a public binary version of our drivers and firmware for the Fetch Research Platforms.
+
+https://fetchrobotics.com/robotics-platforms/
```
Link

erelson replied:
Markdown will render these fine. Is this a style preference?

cjds commented on fetch_binary_drivers/README.md:
```diff
@@ -0,0 +1,43 @@
+
+We're in the process of testing melodic, and setting up our hosted stable mirror, and documenting the upgrade process to ensure Fetch Research Platform customers have a smooth transition.
+
+We will announce to our customers when we're officially ready and supporting melodic.
```
ROS Melodic

cjds commented on fetch_binary_drivers/README.md:
```diff
@@ -0,0 +1,43 @@
+Fetch Research Platform users access to the drivers/firmware faster via the official ros packages.
+
+To create the output of our private package, we have a special build job which runs inside of a docker container to ensure we don't accidentally
+pull in any private dependencies, and also doesn't output any of the additional commercial robot drivers.
```
Do we need to tell them this?

cjds commented on fetch_system_config/debian/fetch-melodic-config.ros.logrotate:
```diff
@@ -0,0 +1,73 @@
+    # Run scripts once per set of logs
+    sharedscripts
+    firstaction
+        . /opt/ros/indigo/setup.sh
```
indigo?

cjds commented on fetch_system_config/debian/fetch-melodic-config.ros.logrotate:
```diff
@@ -0,0 +1,73 @@
+        done
+    endscript
+    postrotate
+        . /opt/ros/indigo/setup.sh
```
indigo?

erelson replied:
Yup, gotta fix this! I left these logrotate rules out initially, then re-added them long after my earlier s/indigo/melodic pass.

cjds commented on fetch_binary_drivers/package.xml:
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>fetch_drivers</name>
+  <version>0.0.0</version>
```
Are we not syncing version numbers with the drivers themselves?

cjds commented on fetch_binary_drivers/package.xml:
```diff
@@ -0,0 +1,66 @@
+
+
+  <!-- Url tags are optional, but multiple are allowed, one per tag -->
+  <url type="website">http://wiki.ros.org/fetch_drivers</url>
```
`https://wiki.ros.org/fetch_drivers` for extra security

cjds commented on fetch_binary_drivers/package.xml:
```diff
@@ -0,0 +1,66 @@
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <maintainer email="amoriarty@fetchrobotics.com">Alexander Moriarty</maintainer>
+  <maintainer email="opensource@fetchrobotics.com">FetchRobotics Open Source Team</maintainer>
```
Fetch Robotics

cjds commented on .gitignore:
```diff
@@ -0,0 +1,14 @@
+*.changes
+fetch-melodic-config
+freight-melodic-config
+files
```
files? what does this do? who uses this folder?

erelson replied:
it's a file created by dpkg-debbuilder or whatever. I'll add a better comment to the .gitignore file

cjds commented on fetch_system_config/debian/fetch-melodic-config.postinst:
```diff
@@ -0,0 +1,138 @@
+            cp /opt/ros/melodic/share/fetch_bringup/launch/fetch.launch /etc/ros/melodic/robot.launch
+        fi
+
+        # TODO this file is missing in 18.04?
```
What's the story with this TODO. Is it done? https://wiki.debian.org/ConfigurePowerButton

erelson replied:
It's a non-impactful TODO that I'm hoping to get to this week.

cjds commented on fetch_system_config/debian/fetch-melodic-config.postinst:
```diff
@@ -0,0 +1,138 @@
+        # Modify default/grub to not use new-style network device naming
+        sed -i "s/GRUB_CMDLINE_LINUX=\"\"/GRUB_CMDLINE_LINUX=\"net.ifnames=0 biosdevname=0\"/" /etc/default/grub
+        update-grub
+        # Determine which device to map to eth0 and eth1 and update /etc/udev/rules.d/70-persistent-net.rules
```
Do eth0 and 1 apply on ubuntu 18.04

erelson replied:
It is for our robots because I found the computer hardware we use is not 100% guaranteed to have the same e.g. ensp1s0 style names every time it boots. Therefore I implement a "standard" way to disable this newer naming behavior.  See the commit message as well.  I'll also being including notes about this networking decision in the docs for 14.04->18.04.

cjds commented on fetch_system_config/debian/fetch-melodic-config.postinst:
```diff
@@ -0,0 +1,138 @@
+                cp -n /etc/udev/rules.d/70-persistent-net.rules /etc/udev/rules.d/70-persistent-net.rules.bak
+            fi
+            cp -f /etc/udev/rules.d/70-persistent-net.rules /tmp/70-persistent-net.rules
+            grep -v -E "eth0|eth1" /tmp/70-persistent-net.rules > /etc/udev/rules.d/70-persistent-net.rules | true
```
eth0 1?

cjds commented on fetch_system_config/root/opt/ros/roscore_poststart.bash:
```diff
@@ -0,0 +1,9 @@
+# This script is used by the roscore systemd service unit
+echo "waiting for roscore to come up"
```
Put the echo in the while so we know it hasn't locked up maybe?

erelson replied:
I'm going to leave this the way it was for simplicity right now.

cjds commented on fetch_binary_drivers/package.xml:
```diff
@@ -0,0 +1,66 @@
+  <exec_depend>fetch_auto_dock_msgs</exec_depend>
+  <exec_depend>nav_msgs</exec_depend>
+  <exec_depend>power_msgs</exec_depend>
+  <exec_depend>robot_calibration_msgs</exec_depend>
```
todo: ensure that these are the same as the ones we used when building the binary drivers.

mikeferguson replied:
you won't be able to do that -- the ROS buildfarm / repos only contain the latest debs - once a new version is synced, the old ones go away. Hardcoding the value would make your package uninstallable at best.

moriarty replied:
@mikeferguson: would a nightly build job of our binaries, ensuring we’ve always got the latest dependencies work here?

@cjds and I were hoping to open source the Dockerfile & Python script which builds the `output.tar.gz`- it won’t be helpful without pull access on our drivers but it would at least make the process transparent

mikeferguson replied:
@moriarty even having the released deb download a nightly binary build is a challenge -- when a sync to public happens, you'd have a ~24 hour window where the version built against public is no longer right. (and that sync happens every ~ 2 weeks, so that's about 7% downtime). Also how do you make it work on both shadow-fixed (where developers tend to test) and the main ros repo?

moriarty replied:
:disappointed:
@dbking77 & @chadrockey maybe long term we reopen the discussion about refactoring some things to better support public/private drivers to make releasing to research customers easier :question: 

We might need a second version setup up and built against shadow-fixed, this still doesn't fix the original problem, and it also wouldn't be clear how to set that up so that the shadow-fixed pulled a `fetch-drivers-shadow.tar.gz` and the release pulled `fetch-drivers.tar.gz` especially since currently they're pulling: `http://packages.fetchrobotics.com/binaries/fetch-drivers-0.8.0.tar.gz`
... Fetch3 & my machine are currently setup against shadow-fixed, but these drivers were built against whatever is inside of

```dockerfile
FROM ros:melodic-robot-bionic
```
(As mentioned @cjds will cleanup the creation process and we plan to put them:
https://github.com/fetchrobotics/fetch_firmware_repackager or rename that repository `fetch_drivers_repackager`)

And it works now, but that's likely because there hasn't been any API/ABI breaking changes, which is not always guaranteed.
We'll need to check which of the dependencies we have are core dependencies, and which are non-core dependencies. 

For the most part, the 7% downtime is a worst case calculation coming from 24hr every 2 weeks. If the dependencies are part of the core ros repositories, I think the API/ABI is stable enough or at least suppose to be... For non-core, and not released by us then it's just a risk that we'll probably need to take.
I don't even know if the melodic sync is triggered every 2 weeks because most of the folks at open robots are developing towards ros2. @jacobperron & @wjwwood?

Our plan, was to setup a mirror of ROS, which we manually trigger a sync on.
So customers with Fetch Research Robots could point to our http://packages.fetchrobotics.com/ and get guaranteed stability. But have the option to get the packages from packages.ros.org either ros or ros-shadow-fixed if they really want to live on the edge and accept some instability.

By pushing this into the public build farm, our internal `packages.fetchrobotics.com` will become almost a pure mirror, which should have much maintainable for our DevOps team and @bluryi.

If we sync our mirror manually, after testing to know we're not in the 7% downtime, then we'll get stability and more frequent updates.

For the 7% of possible down time for the `packages.ros.org` I think @erelson will need to update the (not yet existing) documentation to include strong warnings about pointing to `packages.fetchrobotics.com` or `packages.ros.org`
The packages.fetchrobotics.com mirror for *melodic* does not yet exist, we were going to quietly get our packages into melodic, and then setup a mirror once they were in and synced.

wjwwood replied:
@clalancette is the ROS Boss for Melodic, so it's up to him to decide when the syncs happen, though I think they've been trying to keep a regular cadence. Looking at the discourse posts, it's more like once a month, though maybe that will change.

Perhaps you can anticipate the syncs and be prepared with updated packages of your own to sync in order to minimize the period where your debs are out of date. We do usually announce an intention to sync a week or two before doing it, with the intention that people could test it for unintended regressions.

----

I don't know if this would be relevant as well, but there's this recent announcement by @nuclearsandwich:

https://discourse.ros.org/t/announcing-ros-snapshot-repositories/7705

clalancette replied:
My intention is to sync Melodic every two weeks, though I've basically never been able to achieve that due to regressions that I have to fix for every sync.  So the cadence has indeed been more like once a month.

I intend to continue to attempt to sync every two weeks unless there is a compelling reason to change that.  So if you have reasons you'd like it to change, please let me know and we can consider it.

moriarty replied:
@nickswalker and @mikeferguson I wanted to give you a warning: just in case you've been running melodic and testing with Fetch's using **ros-shadow-fixed** this is now there:

http://repositories.ros.org/status_page/ros_melodic_default.html?q=fetch_

@erelson will do a test on hardware using the **ros-shadow-fixed** before we make an official announcement with links to upgrade documentation. So: best case, everything works, but it's more likely that the official announcement to the fetch user mailing lists is going to be not the upcoming ros melodic sync, but the one after. If we're really unlucky we could be in one of those 7% downtime scenarios which was mentioned above.

erelson commented on fetch_binary_drivers/package.xml:
```diff
@@ -29,9 +29,28 @@
+  <exec_depend>robot_calibration_msgs</exec_depend>
+  <exec_depend>sensor_msgs</exec_depend>
+
+  <!-- ros dependencies wich the drivers dynamically link against -->
```
typo 'wich'

erelson commented on fetch_binary_drivers/package.xml:
```diff
@@ -0,0 +1,66 @@
+    The public fetch_drivers package is a binary only release.
+
+    fetch_drivers contains both the drivers and firmware for the fetch and freight research robots.
+    There should be no reason to use these drivers unless you're running on a fetch or a frieght research robot.
```
typo frieght

erelson reviewed (commented):
Couple typos I found

cjds reviewed (changes requested):
First pass. Will do a more thorough review soon

erelson committed:
Ensure robots power off safely when instructed to

Also:
- Fix remaining indigo references
- Don't worry about conflicting packages that never existed
- String fixes

erelson committed:
Update .gitignore

erelson committed:
Fix typo in .gitignore

erelson committed:
Update wifi power management disabling

erelson commented on fetch_system_config/debian/fetch-melodic-config.postinst:
```diff
@@ -0,0 +1,141 @@
+
+        # Disable wifi power management in 18.04
+        # https://gist.github.com/jcberthon/ea8cfe278998968ba7c5a95344bc8b55
+        sed -i "s/wifi.powersave = 3/wifi.powersave = 2/" /etc/NetworkManager/conf.d/default-wifi-powersave-on.conf
```
fyi @cjds here's the updated approach :+1:

erelson committed:
Copy updates to freight post-install script

cjds reviewed (approved):
LGTM. This works for a 0.1 release. We'll be improving it over time.

moriarty commented:
Once https://github.com/ros/rosdistro/pull/20301 is in the rosdep keys for `fetch_description` and `fetch_teleop` should be available and then I can bloom this.

moriarty commented:
https://github.com/ros/rosdistro/pull/20303

@cjds and @erelson we'll test this once on hardware before we merge the PR to the documentation website which officially announces melodic upgrade process.
